go/fixchain/logger.go

```diff
@@ -182,7 +174,35 @@ func (l *Logger) postChain(p *toPost) {
 			} else {
 				log.Printf(ferr.Error.Error())
 				p.retries--
-				l.postToLog(p)
+				// asyncPostToLog() is called instead of its synchronous
+				// counterpart, postToLog(), to avoid deadlock during retries.
+				// Without the separate goroutine created in asyncPostToLog(),
+				// deadlock can occur in the following situation:
+				//
+				// Suppose there is only one postServer() goroutine running, and
+				// it is blocked waiting for a toPost on the toPost chan.  A
+				// toPost gets added to the chan, which causes the following to
+				// happen:
+				// - the postServer takes the toPost from the chan.
+				// - the postServer calls l.postChain(toPost), and waits for
+				//   l.postChain() to return before going back to the toPost
+				//   chan for another toPost.
+				// - l.postChain() begins execution.  Suppose the first post
+				//   attempt of the toPost fails for some network-related
+				//   reason.
+				// - l.postChain retries and calls l.postToLog() to queue up the
+				//   toPost to try to post it again.
+				// - l.postToLog() tries to put the toPost on the toPost chan,
+				//   and blocks until a postServer takes it off the chan.
+				// But the one and only postServer is still waiting for
+				// l.postChain (and therefore l.postToLog) to return, and will
+				// not go to take another toPost off the toPost chan until that
+				// happens.
+				// Thus, deadlock.
+				//
+				// Similar situations with multiple postServers can easily be
+				// imagined.
```

Comment:
Okay.  Should the comment in QueueChain() about why we're using postToLog instead of asyncPostToLog in that situation be moved to postToLog instead too then?  It's the same type of comment, but with the functions reversed, and a different reason.